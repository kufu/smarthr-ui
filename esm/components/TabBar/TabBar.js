var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import React from 'react';
import styled, { css } from 'styled-components';
import { useTheme } from '../../hooks/useTheme';
export var TabBar = function (_a) {
    var _b = _a.className, className = _b === void 0 ? '' : _b, _c = _a.bordered, bordered = _c === void 0 ? true : _c, children = _a.children;
    var theme = useTheme();
    var classNames = className + " " + (bordered ? 'bordered' : '');
    return (React.createElement(Wrapper, { role: "tablist", className: classNames, themes: theme }, children));
};
var Wrapper = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", "\n"], ["\n  ",
    "\n"])), function (_a) {
    var themes = _a.themes;
    var border = themes.frame.border;
    return css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      display: flex;\n\n      &.bordered {\n        position: relative;\n\n        :after {\n          position: absolute;\n          left: 0;\n          right: 0;\n          bottom: 0px;\n          border-bottom: ", ";\n          content: '';\n        }\n\n        > button {\n          &.selected {\n            position: relative;\n            z-index: 1;\n          }\n        }\n      }\n    "], ["\n      display: flex;\n\n      &.bordered {\n        position: relative;\n\n        :after {\n          position: absolute;\n          left: 0;\n          right: 0;\n          bottom: 0px;\n          border-bottom: ", ";\n          content: '';\n        }\n\n        > button {\n          &.selected {\n            position: relative;\n            z-index: 1;\n          }\n        }\n      }\n    "])), border.default);
});
var templateObject_1, templateObject_2;
//# sourceMappingURL=TabBar.js.map