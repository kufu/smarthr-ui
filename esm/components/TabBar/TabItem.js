var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import React from 'react';
import styled, { css } from 'styled-components';
import { useTheme } from '../../hooks/useTheme';
import { isTouchDevice } from '../../libs/ua';
export var TabItem = function (_a) {
    var id = _a.id, children = _a.children, onClick = _a.onClick, _b = _a.selected, selected = _b === void 0 ? false : _b, _c = _a.className, className = _c === void 0 ? '' : _c, _d = _a.disabled, disabled = _d === void 0 ? false : _d;
    var theme = useTheme();
    var classNames = className + " " + (selected ? 'selected' : '');
    return (React.createElement(Wrapper, { role: "tab", "aria-selected": selected, className: classNames, onClick: function () { return onClick(id); }, disabled: disabled, themes: theme }, children));
};
var resetButtonStyle = css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  background-color: transparent;\n  border: none;\n  cursor: pointer;\n  outline: none;\n  padding: 0;\n  appearance: none;\n"], ["\n  background-color: transparent;\n  border: none;\n  cursor: pointer;\n  outline: none;\n  padding: 0;\n  appearance: none;\n"])));
var Wrapper = styled.button(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n  ", "\n"], ["\n  ", "\n  ",
    "\n"])), resetButtonStyle, function (_a) {
    var themes = _a.themes;
    var size = themes.size, palette = themes.palette, interaction = themes.interaction;
    return css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n      font-weight: bold;\n      font-size: ", ";\n      color: ", ";\n      height: 40px;\n      border-bottom: solid 3px transparent;\n      padding: 0 ", ";\n      box-sizing: border-box;\n      transition: ", ";\n\n      &.selected {\n        color: ", ";\n        border-color: ", ";\n      }\n\n      :hover {\n        background-color: ", ";\n        color: ", ";\n      }\n\n      :disabled {\n        background-color: transparent;\n        color: ", ";\n        cursor: not-allowed;\n      }\n    "], ["\n      font-weight: bold;\n      font-size: ", ";\n      color: ", ";\n      height: 40px;\n      border-bottom: solid 3px transparent;\n      padding: 0 ", ";\n      box-sizing: border-box;\n      transition: ",
        ";\n\n      &.selected {\n        color: ", ";\n        border-color: ", ";\n      }\n\n      :hover {\n        background-color: ", ";\n        color: ", ";\n      }\n\n      :disabled {\n        background-color: transparent;\n        color: ", ";\n        cursor: not-allowed;\n      }\n    "])), size.pxToRem(size.font.TALL), palette.TEXT_GREY, size.pxToRem(size.space.S), isTouchDevice
        ? 'none'
        : "background-color " + interaction.hover.animation + ", color " + interaction.hover.animation, palette.TEXT_BLACK, palette.MAIN, palette.COLUMN, palette.TEXT_BLACK, palette.disableColor(palette.TEXT_GREY));
});
var templateObject_1, templateObject_2, templateObject_3;
//# sourceMappingURL=TabItem.js.map