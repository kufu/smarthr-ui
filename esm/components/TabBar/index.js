export { TabBar } from './TabBar';
export { TabItem } from './TabItem';
//# sourceMappingURL=index.js.map