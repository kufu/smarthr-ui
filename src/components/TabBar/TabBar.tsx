import React, { FC, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { Theme, useTheme } from '../../hooks/useTheme'

type Props = {
  children: ReactNode
  bordered?: boolean
  className?: string
}

export const TabBar: FC<Props> = ({ className = '', bordered = true, children }) => {
  const theme = useTheme()
  const classNames = `${className} ${bordered ? 'bordered' : ''}`

  return (
    <Wrapper role="tablist" className={classNames} themes={theme}>
      {children}
    </Wrapper>
  )
}

const Wrapper = styled.div<{ themes: Theme }>`
  ${({ themes }) => {
    const { border } = themes.frame

    return css`
      display: flex;
      background-color: #fff;

      &.bordered {
        position: relative;

        :after {
          position: absolute;
          bottom: 0px;
          width: 100%;
          border-bottom: ${border.default};
          content: '';
        }

        > button {
          &.selected {
            position: relative;
            z-index: 1;
          }
        }
      }
    `
  }}
`
