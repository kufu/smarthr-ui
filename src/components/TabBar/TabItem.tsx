import React, { FC, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { Theme, useTheme } from '../../hooks/useTheme'
import { isTouchDevice } from '../../libs/ua'

type Props = {
  id: string
  children: ReactNode
  selected?: boolean
  disabled?: boolean
  className?: string
  onClick: (tabId: string) => void
}

export const TabItem: FC<Props> = ({
  id,
  children,
  onClick,
  selected = false,
  className = '',
  disabled = false,
}) => {
  const theme = useTheme()
  const classNames = `${className} ${selected ? 'selected' : ''}`

  return (
    <Wrapper
      role="tab"
      aria-selected={selected}
      className={classNames}
      onClick={() => onClick(id)}
      disabled={disabled}
      themes={theme}
    >
      {children}
    </Wrapper>
  )
}

const resetButtonStyle = css`
  background-color: transparent;
  border: none;
  cursor: pointer;
  outline: none;
  padding: 0;
  appearance: none;
`
const Wrapper = styled.button<{ themes: Theme }>`
  ${resetButtonStyle}
  ${({ themes }) => {
    const { size, palette, interaction } = themes
    return css`
      font-weight: bold;
      font-size: ${size.pxToRem(size.font.TALL)};
      color: ${palette.TEXT_GREY};
      height: 40px;
      border-bottom: solid 3px transparent;
      padding: 0 ${size.pxToRem(size.space.S)};
      box-sizing: border-box;
      transition: ${isTouchDevice
        ? 'none'
        : `background-color ${interaction.hover.animation}, color ${interaction.hover.animation}`};

      &.selected {
        color: ${palette.TEXT_BLACK};
        border-color: ${palette.MAIN};
      }

      :hover {
        background-color: ${palette.COLUMN};
        color: ${palette.TEXT_BLACK};
      }

      :disabled {
        background-color: inherit;
        color: ${palette.disableColor(palette.TEXT_GREY)};
        cursor: not-allowed;
      }
    `
  }}
`
