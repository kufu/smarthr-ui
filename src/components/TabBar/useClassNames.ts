import { useMemo } from 'react'

import { useClassNameGenerator } from '../../hooks/useClassNameGenerator'
import { TabBar, TabItem } from './'

export function useClassNames() {
  const generate = useClassNameGenerator(TabBar.displayName || 'TabBar')
  const generateForItem = useClassNameGenerator(TabItem.displayName || 'TabItem')
  return useMemo(
    () => ({
      tabBar: {
        wrapper: generate(),
      },
      tabItem: {
        wrapper: generateForItem(),
      },
    }),
    [generate, generateForItem],
  )
}
