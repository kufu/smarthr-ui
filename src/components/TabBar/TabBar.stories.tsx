import * as React from 'react'
import { action } from '@storybook/addon-actions'
import { Story } from '@storybook/react'
import { userEvent } from '@storybook/testing-library'

import styled from 'styled-components'

import { TabItem } from './TabItem'
import { TabBar } from './TabBar'

export default {
  title: 'TabBar',
  component: TabBar,
  subcomponents: { TabItem },
}

const Template: Story = () => (
  <Ul>
    <li>
      <p>Border</p>
      <TabBar>
        <TabItem id="1" onClick={action('clicked')}>
          Tab
        </TabItem>
        <TabItem id="2" onClick={action('clicked')} selected>
          Selected
        </TabItem>
        <TabItem id="3" onClick={action('clicked')} disabled>
          Disabled
        </TabItem>
      </TabBar>
    </li>
    <li>
      <p>No border</p>
      <TabBar bordered={false}>
        <TabItem id="1" onClick={action('clicked')}>
          Tab
        </TabItem>
        <TabItem id="2" onClick={action('clicked')} selected>
          Selected
        </TabItem>
        <TabItem id="3" onClick={action('clicked')} disabled>
          Disabled
        </TabItem>
      </TabBar>
    </li>
  </Ul>
)

export const All = Template.bind({})

export const RegFocusBorder = Template.bind({})
RegFocusBorder.play = () => userEvent.tab()

export const RegFocusNoBorder = Template.bind({})
RegFocusNoBorder.play = () => [...Array(3)].forEach((_) => userEvent.tab())

const Ul = styled.ul`
  padding: 0 1rem;

  li {
    margin-bottom: 1rem;
    list-style: none;
  }
`
