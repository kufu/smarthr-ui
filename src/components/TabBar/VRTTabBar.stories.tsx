import { action } from '@storybook/addon-actions'
import { StoryFn } from '@storybook/react'
import { fireEvent, within } from '@storybook/testing-library'
import React from 'react'
import styled from 'styled-components'

import { InformationPanel } from '../InformationPanel'

import { TabBar } from './TabBar'
import { All } from './TabBar.stories'
import { TabItem } from './TabItem'

export default {
  title: 'Navigation（ナビゲーション）/TabBar',
  component: TabBar,
  parameters: {
    withTheming: true,
  },
}

export const VRTState: StoryFn = () => (
  <>
    <VRTInformationPanel title="VRT 用の Story です" togglable={false}>
      hover, activeなどの状態で表示されます{' '}
    </VRTInformationPanel>
    <Ul>
      <li>
        <p>hover</p>
        <TabBar id="hover">
          <TabItem id="border-1" onClick={action('clicked')}>
            Tab
          </TabItem>
          <TabItem id="border-2" onClick={action('clicked')} selected>
            Selected
          </TabItem>
          <TabItem id="border-3" onClick={action('clicked')} disabled>
            Disabled
          </TabItem>
        </TabBar>
      </li>
      <li>
        <p>focus</p>
        <TabBar id="focus">
          <TabItem id="border-4" onClick={action('clicked')}>
            Tab
          </TabItem>
          <TabItem id="border-5" onClick={action('clicked')} selected>
            Selected
          </TabItem>
          <TabItem id="border-6" onClick={action('clicked')} disabled>
            Disabled
          </TabItem>
        </TabBar>
      </li>
      <li>
        <p>focusVisible</p>
        <TabBar id="focus-visible">
          <TabItem id="border-7" onClick={action('clicked')}>
            Tab
          </TabItem>
          <TabItem id="border-8" onClick={action('clicked')} selected>
            Selected
          </TabItem>
          <TabItem id="border-9" onClick={action('clicked')} disabled>
            Disabled
          </TabItem>
        </TabBar>
      </li>
      <li>
        <p>active</p>
        <TabBar id="active">
          <TabItem id="border-10" onClick={action('clicked')}>
            Tab
          </TabItem>
          <TabItem id="border-11" onClick={action('clicked')} selected>
            Selected
          </TabItem>
          <TabItem id="border-12" onClick={action('clicked')} disabled>
            Disabled
          </TabItem>
        </TabBar>
      </li>
    </Ul>
  </>
)
VRTState.parameters = {
  controls: { hideNoControlsWarning: true },
  pseudo: {
    hover: ['#hover button'],
    focus: ['#focus button'],
    focusVisible: ['#focus-visible button'],
    active: ['#active button'],
  },
}

export const VRTScroll: StoryFn = () => (
  <>
    <VRTInformationPanel title="VRT 用の Story です" togglable={false}>
      画面幅が狭い状態でスクロールされるか確認します
    </VRTInformationPanel>
    <Ul>
      <li>
        <TabBar>
          <TabItem id="border-1" onClick={action('clicked')}>
            Tab1
          </TabItem>
          <TabItem id="border-2" onClick={action('clicked')}>
            Tab2
          </TabItem>
          <TabItem id="border-3" onClick={action('clicked')}>
            Tab3
          </TabItem>
          <TabItem id="border-4" onClick={action('clicked')}>
            Tab4
          </TabItem>
          <TabItem id="border-5" onClick={action('clicked')}>
            Tab5
          </TabItem>
          <TabItem id="border-6" onClick={action('clicked')}>
            Tab6
          </TabItem>
        </TabBar>
      </li>
    </Ul>
  </>
)
VRTScroll.parameters = {
  viewport: {
    defaultViewport: 'vrtMobile',
  },
  chromatic: {
    modes: {
      vrtMobile: { viewport: 'vrtMobile' },
    },
  },
}
VRTScroll.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement)
  // nav要素の直下のdiv要素がスクロールする要素
  const scrollableElement = await canvas.findByRole('tablist')
  await fireEvent.scroll(scrollableElement, { target: { scrollLeft: 1000 } })
}

export const VRTForcedColors: StoryFn = () => (
  <>
    <VRTInformationPanel title="VRT 用の Story です" togglable={false}>
      Chromatic 上では強制カラーモードで表示されます
    </VRTInformationPanel>
    <All />
  </>
)
VRTForcedColors.parameters = {
  chromatic: { forcedColors: 'active' },
}

const VRTInformationPanel = styled(InformationPanel)`
  margin-bottom: 24px;
`

const Ul = styled.ul`
  padding: 0 1rem;

  li {
    margin-bottom: 1rem;
    list-style: none;
  }
`
