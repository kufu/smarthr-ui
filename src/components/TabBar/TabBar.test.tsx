import { create } from 'react-test-renderer'
import React from 'react'

import { TabBar } from './TabBar'
import { TabItem } from './TabItem'

const Tab = () => (
  <TabBar>
    <TabItem id="1" onClick={jest.fn()}>
      Tab
    </TabItem>
    <TabItem id="2" onClick={jest.fn()} selected>
      Selected
    </TabItem>
    <TabItem id="3" onClick={jest.fn()} disabled>
      Disabled
    </TabItem>
  </TabBar>
)

describe('TabBar', () => {
  it('should be match snapshot', () => {
    const testRenderer = create(<Tab />)
    expect(testRenderer.toJSON()).toMatchSnapshot()
  })
})
