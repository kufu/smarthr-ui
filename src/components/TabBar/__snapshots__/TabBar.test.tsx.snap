// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar should be match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0.bordered {
  position: relative;
}

.c0.bordered::before {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  border-bottom: 1px solid #d6d3d0;
  content: '';
}

.c1 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  padding: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-weight: bold;
  font-size: 0.875rem;
  color: #706d65;
  height: 40px;
  border-bottom: solid 3px transparent;
  padding: 0 1.5rem;
  box-sizing: border-box;
  -webkit-transition: background-color .3s ease-out,color .3s ease-out;
  transition: background-color .3s ease-out,color .3s ease-out;
}

.c1.selected {
  position: relative;
  color: #23221f;
  border-color: #0077c7;
}

.c1:hover {
  background-color: #f8f7f6;
  color: #23221f;
}

.c1:disabled {
  background-color: transparent;
  color: rgba(112,109,101,0.5);
  cursor: not-allowed;
}

<div
  className="c0  bordered"
  role="tablist"
>
  <button
    aria-selected={false}
    className="c1  "
    disabled={false}
    onClick={[Function]}
    role="tab"
    type="button"
  >
    Tab
  </button>
  <button
    aria-selected={true}
    className="c1  selected"
    disabled={false}
    onClick={[Function]}
    role="tab"
    type="button"
  >
    Selected
  </button>
  <button
    aria-selected={false}
    className="c1  "
    disabled={true}
    onClick={[Function]}
    role="tab"
    type="button"
  >
    Disabled
  </button>
</div>
`;
